Feature request: switch to work with student repositories named in GitHub Classroom style

**Feature Request:**

I've been using GitHub Classroom, which automatically creates student repositories named in this style:  `<assignment-name>-<username>`.

It would be nice to have a command-line switch or setting that would make it easy to use student repositories named in the style `<assignment-name>-<username>` instead of the default `<username>-<assignment-name>` style.

Looks like I could change this by a simple edit to the `generate_repo_name` function in the `name.py` file to this line:

```
return "{}-{}".format(team_name, assignment_name)
```

Thank you.

The testing I did above used RepoBee-deployed test repositories that already had the team direct accesses auto-setup.  

Now if I try to do `reviews assign` on a set of previously-deployed GitHub Classroom repositories, it fails because the team direct write access has not been set:

```
$ repobee reviews assign -a 3-r-characters-homework --sf dw_students.txt --num-reviews 1 --issue review_issue.md
Fetching teams and repos: 100%| 2/2 [00:00<00:00,  2.28it/s]
[ERROR] NotFoundError: Can't find repos: 3-r-characters-homework-DanielEWeeks, 3-r-characters-homework-daniel-e-weeks
[ERROR] RepoBee exited unexpectedly. Please visit the FAQ to try to resolve the problem: https://repobee.readthedocs.io/en/stable/faq.html
```

The GItHub Classroom-created repositories `3-r-characters-homework-DanielEWeeks` and `3-r-characters-homework-daniel-e-weeks` exist, but the `DanielEWeeks` team and the `daniel-e-weeks` team did not have direct write access to their own homework repositories.

After adding the required team direct accesses, then the `review assign` command worked.

So if someone has already deployed a GitHub Classroom assignment, to use RepoBee with that set of assignments, one needs to do a `repos setup` type of command where it doesn't create the already existing repositories, but assigns the teams.

Is it save to apply 'repos setup' on an already deployed set of repositories that the students have already worked on?  That's not clear from the documentation - wouldn't want to inadvertently wipe out students' work in their repositories by mistake.

--- a/src/repobee_plug/__init__.py
+++ b/src/repobee_plug/__init__.py
@@ -1,5 +1,3 @@
-import pluggy  # type: ignore
-
 from repobee_plug.__version import __version__  # noqa: F401
 
 from repobee_plug import types
@@ -59,10 +57,10 @@ from repobee_plug.exceptions import (
 from repobee_plug.localreps import StudentTeam, StudentRepo, TemplateRepo
 
 # Hook functions
+from repobee_plug.hookmanager import manager
 import repobee_plug._corehooks
 import repobee_plug._exthooks
 
-manager = pluggy.PluginManager(__package__)
 manager.add_hookspecs(repobee_plug._corehooks)
 manager.add_hookspecs(repobee_plug._exthooks)
 

--- a/src/repobee_plug/_corehooks.py
+++ b/src/repobee_plug/_corehooks.py
@@ -9,9 +9,9 @@ to allow for this dynamic override.
     :synopsis: Hookspecs for repobee core hooks.
 """
 
-from typing import List, Tuple
+from typing import List, Tuple, Union
 
-from repobee_plug import localreps, hook, reviews
+from repobee_plug import localreps, hook, reviews, platform
 
 
 #####################
@@ -89,4 +89,28 @@ def api_init_requires() -> Tuple[str]:
 
     Returns:
         Names of the required arguments.
+    """
+
+
+############################
+# Hooks for naming schemes #
+############################
+
+
+@hook.hookspec(firstresult=True)
+def generate_repo_name(
+    team_name: Union[str, localreps.StudentTeam, platform.Team],
+    assignment_name: str,
+) -> str:
+    """This hook allows for overriding the behavior of
+    :py:func:`repobee_plug.name.generate_repo_name`.
+
+    .. danger::
+
+        The implementations of this hook should never be invoked other than in
+        :py:func:`repobee_plug.name.generate_repo_name`.
+
+    Args:
+        team_name: Name of the associated team.
+        assignment_name: Name of an assignment.
     """

--- a/src/repobee_plug/hookmanager.py
+++ b/src/repobee_plug/hookmanager.py
@@ -1,0 +1,4 @@
+"""Module solely for the hook manager"""
+import pluggy  # type: ignore
+
+manager = pluggy.PluginManager(__package__)

--- a/src/repobee_plug/name.py
+++ b/src/repobee_plug/name.py
@@ -7,6 +7,7 @@ from typing import Iterable, Union
 
 from repobee_plug import localreps
 from repobee_plug import platform
+from repobee_plug.hookmanager import manager
 
 
 def generate_repo_names(
@@ -39,11 +40,16 @@ def generate_repo_name(
 ) -> str:
     """Construct a repo name for a team.
 
+    The behavior of this function can be overridden by implementing the
+    :py:func:`repobee_plug._corehooks.generate_repo_name` hook.
+
     Args:
         team_name: Name of the associated team.
         assignment_name: Name of an assignment.
     """
-    return "{}-{}".format(team_name, assignment_name)
+    return manager.hook.generate_repo_name(
+        team_name=team_name, assignment_name=assignment_name
+    ) or "{}-{}".format(team_name, assignment_name)
 
 
 def generate_review_team_name(

--- a/tests/new_integration_tests/test_repos.py
+++ b/tests/new_integration_tests/test_repos.py
@@ -1,6 +1,7 @@
 """Tests for the repos category of commands."""
 import pathlib
 import tempfile
+import itertools
 
 from typing import List, Mapping, Tuple, Iterable
 
@@ -207,7 +208,7 @@ class TestSetup:
 
         # nothing should have changed, and there should be no errors
         assert_student_repos_match_templates(
-            STUDENT_TEAMS, TEMPLATE_REPO_NAMES, funcs.get_repos(platform_url),
+            STUDENT_TEAMS, TEMPLATE_REPO_NAMES, funcs.get_repos(platform_url)
         )
         assert "[ERROR]" not in capsys.readouterr().out
 
@@ -279,6 +280,36 @@ class TestSetup:
 
         assert not expected_repo_names
 
+    def test_use_non_standard_repo_names(self, platform_url):
+        """Test setting up repos with non-standard repo names using an
+        implementation of the ``generate_repo_name`` hook.
+        """
+
+        def generate_repo_name(team_name, assignment_name):
+            return f"{assignment_name}-BONKERS-{team_name}"
+
+        expected_repo_names = [
+            generate_repo_name(team, assignment_name)
+            for team, assignment_name in itertools.product(
+                const.STUDENT_TEAMS, const.TEMPLATE_REPO_NAMES
+            )
+        ]
+
+        class StrangeNamingConvention(plug.Plugin):
+            def generate_repo_name(self, team_name, assignment_name):
+                return generate_repo_name(team_name, assignment_name)
+
+        funcs.run_repobee(
+            f"repos setup -a {const.TEMPLATE_REPOS_ARG} "
+            f"--base-url {platform_url}",
+            plugins=[StrangeNamingConvention],
+        )
+
+        actual_repo_names = [
+            repo.name for repo in funcs.get_repos(platform_url)
+        ]
+        assert sorted(actual_repo_names) == sorted(expected_repo_names)
+
 
 class TestClone:
     """Tests for the ``repos clone`` command."""
@@ -294,6 +325,48 @@ class TestClone:
             assert_cloned_student_repos_match_templates(
                 STUDENT_TEAMS, TEMPLATE_REPO_NAMES, workdir
             )
+
+    def test_use_non_standard_repo_names(self, platform_url, tmp_path_factory):
+        """Test cloning repos with non-standard repo names using an
+        implementation of the ``generate_repo_name`` hook.
+        """
+        # arrange
+        def generate_repo_name(team_name, assignment_name):
+            return f"{assignment_name}-BONKERS-{team_name}"
+
+        class StrangeNamingConvention(plug.Plugin):
+            def generate_repo_name(self, team_name, assignment_name):
+                return generate_repo_name(team_name, assignment_name)
+
+        workdir = tmp_path_factory.mktemp("workdir")
+        expected_repo_names = [
+            generate_repo_name(team, assignment_name)
+            for team, assignment_name in itertools.product(
+                const.STUDENT_TEAMS, const.TEMPLATE_REPO_NAMES
+            )
+        ]
+        funcs.run_repobee(
+            f"repos setup -a {const.TEMPLATE_REPOS_ARG} "
+            f"--base-url {platform_url}",
+            plugins=[StrangeNamingConvention],
+        )
+
+        # act
+        funcs.run_repobee(
+            f"repos clone -a {const.TEMPLATE_REPOS_ARG} "
+            f"--base-url {platform_url}",
+            workdir=workdir,
+            plugins=[StrangeNamingConvention],
+        )
+
+        # assert
+        local_repos = itertools.chain.from_iterable(
+            student_dir.iterdir() for student_dir in workdir.iterdir()
+        )
+        actual_repo_names = [
+            repo_path.name for repo_path in local_repos if repo_path.is_dir()
+        ]
+        assert sorted(actual_repo_names) == sorted(expected_repo_names)
 
     def test_clone_discover_repos(self, platform_url, with_student_repos):
         with tempfile.TemporaryDirectory() as tmp:

